include/swift/Driver/ExperimentalDependencyDriverGraph.h

```diff
@@ -0,0 +1,338 @@
+  /// file whose swiftdeps has not been read yet. For these, the filename is the
+  /// empty string.
+  ///
+  /// Don't add to this collection directly; use \ref addToMap
+  /// instead because it enforces the correspondence with the swiftFileDeps
+  /// field of the node.
+  /// TODO: Fix above comment
+  ///
+  /// Sadly, cannot use an optional string for a key.
+  using NodeMap =
+      BiIndexedTwoStageMap<std::string, DependencyKey, DriverNode *>;
+  NodeMap nodeMap;
+
+  /// Since dependency keys use baseNames, they are coarser than individual
+  /// decls. So two decls might map to the same key. Index the second level by
+  /// file. Very useful to find a Node given the key, for instance when finding
+  /// all the nodes affected by a change in one.
+
+  /// Tracks def-use relationships by DependencyKey.
+  std::unordered_map<DependencyKey, std::unordered_set<DependencyKey>>
+      usesByDef;
+
+  // Supports requests from the driver to getExternalDependencies.
+  std::unordered_set<std::string> externalDependencies;
+
+  /// The new version of "Marked."
+  /// Record cascading jobs by swiftDepsFilename because that's what
+  /// nodes store directly.
+  ///
+  /// The status quo system uses "cascade" for the following:
+  /// Def1 -> def2 -> def3, where arrows are uses, so 3 depends on 2 which
+  /// depends on 1. The first use is said to "cascade" if when def1 changes,
+  /// def3 is dirtied.
+  /// TODO: Move cascadingJobs out of the graph, ultimately.
+  /// If marked, any Job that depends on me must be rebuilt after compiling me
+  /// if I have changed.
+
+  std::unordered_set<std::string> cascadingJobs;
+
+  /// Keyed by swiftdeps filename, so we can get back to Jobs.
+  std::unordered_map<std::string, const Job *> jobsBySwiftDeps;
+
+  /// For debugging, the driver writes out a dot file of the graph every time a
+  /// Frontend swiftdeps is read and integrated. In order to keep subsequent
+  /// files for the same job distinct, keep a sequence number for each job.
+  std::unordered_map<const Job *, uint> dotFileSequenceNumberByJob;
+
+  /// Encapsulate the invariant between where the node resides in
+  /// nodesBySwiftDepsFile and the swiftDeps node instance variable here.
+  void addToMap(DriverNode *n) {
+    nodeMap.insert(n->getSwiftDeps().getValueOr(std::string()), n->getKey(), n);
+  }
+
+  /// When integrating a FrontendGraph, there might be a node representing a
+  /// Decl that had previously been read as an expat, that is a node
+  /// representing a Decl in no known file (to that point). (Recall the the
+  /// Frontend processes name lookups as dependencies, but does not record in
+  /// which file the name was found.) In such a case, it is necessary to move
+  /// the node to the proper collection.
+  void moveNodeToDifferentFile(DriverNode *n, Optional<std::string> newFile) {
+    eraseNodeFromMap(n);
+    n->setSwiftDeps(newFile);
+    addToMap(n);
+  }
```

Comment:
Hm. In the classic model, two files might `provide` the same base name that a third file `depends` on. How does that work here?

Reply:
When integrating the `FrontendGraph` into the `DriverGraph`, `integrateFrontendNode` is called. If the `provide` is new, `preexistingNodeInPlace` will be `None`, as will be `preexistingExpat` if the third file's graph has not yet been integrated. `dupsExistInOtherFiles` will be true when the second file's graph is integrated. `integrateFrontendDeclNode` will be called and will fall through to `integrateByCreatingANewNode`. The new node gets added to `nodeMap`. Because it is in a different file, it does not replace the first file's node there.

Later on, `checkTransitiveClosureForCascading` calls `forEachUseOf` which will find the use, no matter which of the `provides` has changed, because it is based on keys.

I hope that answers the question:)